src/main/java/com/minecolonies/entity/ai/basic/AbstractEntityAIBasic.java

```diff
@@ -891,24 +1083,118 @@ private void updateToolFlag(@NotNull final String tool, final int required)
      */
     private int getMostEfficientTool(@NotNull final Block target)
     {
-        String tool = target.getHarvestTool(target.getDefaultState());
-        int required = target.getHarvestLevel(target.getDefaultState());
+
+        final String tool = target.getHarvestTool(target.getDefaultState());
+        final int required = target.getHarvestLevel(target.getDefaultState());
         int bestSlot = -1;
         int bestLevel = Integer.MAX_VALUE;
         @NotNull InventoryCitizen inventory = worker.getInventoryCitizen();
+        boolean printMessage = false;
+
+
         for (int i = 0; i < inventory.getSizeInventory(); i++)
         {
-            ItemStack item = inventory.getStackInSlot(i);
-            int level = Utils.getMiningLevel(item, tool);
+            final ItemStack item = inventory.getStackInSlot(i);
+            final int level = Utils.getMiningLevel(item, tool);
+
             if (level >= required && level < bestLevel)
             {
-                bestSlot = i;
-                bestLevel = level;
+                if(item!=null && tool != "")
+                {
+                    final boolean toolVerify = verifyToolLevel2(item, level);
+                    if (toolVerify)
+                    {
+                        bestSlot = i;
+                        bestLevel = level;
+                    }
+
+                }
+                else
+                {
+                    bestSlot = i;
+                    bestLevel = level;
+                }
             }
         }
+
         return bestSlot;
     }
 
+
+    /**
+     * Verifies if there is one tool with an acceptable level
+     * in a worker's inventory
+     *
+     * @param tool the type of tool needed
+     * @return true if tool is acceptable
+     */
+
+    private boolean verifyToolLevel(final String tool)
+    {
+        @NotNull InventoryCitizen inventory = worker.getInventoryCitizen();
+
+
+        for (int i = 0; i < inventory.getSizeInventory(); i++)
+        {
+
+            final ItemStack item = inventory.getStackInSlot(i);
+            final int level = Utils.getMiningLevel(item, tool);
+
+            if(item!=null)
+            {
+                final boolean isTool = Utils.isTool(item,tool);
+                if (isTool && verifyToolLevel2(item, level))
+                    return true;
+            }
+        }
+        return false;
+    }
+
+
+    /**
+     * Verifies if an item has an appropriated grade
+     *
+     * @param item the type of tool needed
+     * @param level the type of tool needed
+     * @return true if tool is acceptable
+     */
+
+
+    private boolean verifyToolLevel2(final ItemStack item, int level)
```

Comment:
why name this 2 what does that mean? Why is the name different? And the case is still weird.